ci/main.cue

```diff
@@ -2,98 +2,148 @@ package main
 
 import (
 	"dagger.io/dagger"
-	"dagger.io/dagger/core"
+
 	"universe.dagger.io/bash"
+	"universe.dagger.io/alpine"
+	"universe.dagger.io/docker"
+	"universe.dagger.io/go"
+
+	"github.com/dagger/dagger/ci/pkg/golangci"
+	"github.com/dagger/dagger/ci/pkg/shellcheck"
+	"github.com/dagger/dagger/ci/pkg/markdownlint"
 )
 
 dagger.#Plan & {
-
 	// FIXME: Ideally we would want to automatically set the platform's arch identical to the host
 	// to avoid the performance hit caused by qemu (linter goes from <3s to >3m when arch is x86)
 	// Uncomment if running locally on Mac M1 to bypass qemu
 	// platform: "linux/aarch64"
 	platform: "linux/amd64"
 
-	client: filesystem: "./build": write: contents: actions.build.export.directories["/build"]
+	client: filesystem: "../": read: exclude: [
+		"ci",
+		"**/node_modules",
+		"cmd/dagger/dagger",
+		"cmd/dagger/dagger-debug",
```

Comment:
This is an anti-pattern: it causes `dagger do` to break when workdir is not `ci`.

Reply:
Yes, but what's the alternative?

The `_source` workaround before -- I lost a good chunk of time figuring out where `_source` came from. It doesn't feel natural to me to have a global variable declared in a parent file, can't think of another programming language using that pattern.

Also -- that workaround only works for `client: filesystem`. As soon as I add integration tests next, the `sops` command in `client: commands` will break when workdir is not `ci` (`sops -d ../tests/secrets_sops.yaml`).

Reply:
> Yes, but what's the alternative?
> The `_source` workaround before -- I lost a good chunk of time figuring out where `_source` came from. It doesn't feel natural to me to have a global variable declared in a parent file, can't think of another programming language using that pattern.

That workaround is definitely not perfect, but it's better than `client.filesystem.".."`. We should keep the least bad workaround while we search for something better.

> Also -- that workaround only works for `client: filesystem`. As soon as I add integration tests next, the `sops` command in `client: commands` will break when workdir is not `ci` (`sops -d ../tests/secrets_sops.yaml`).

Good question. Off the top of my head, why not run sops in a regular action, and give it access to the secrets file via `core.#Source`? If the secrets file is embedded alongside the plan CUE files (which in this case it is), then that's even better since it reduces a dependency on a client command.

Contrast this with eg. `aws-vault` or `docker login` which depend on configuration that is out of scope from the plan: `~/.aws`, `~/.docker`, and even eg. Mac keychain data.. In scenarios like these then `client.command` is a must-have.

Reply:
Zooming out for a second, to question the parameters of our debate.

* These workarounds are only necessary because we moved the plan in `ci`
* We want to replace our  Makefile with a Dagger plan
* It would be awkward if our Makefile were in a `./ci` sub-directory
* Likewise, we should probably move the plan to the top-level, similar to our Makefile, so that `dagger do` works out of the box when developing.
* Nice side-effect: we don't need to argue about these workarounds anymore :)

Reply:
Ok done -- there's a top-level `ci.cue` file now, and the `ci/` directory only contains a few packages (didn't want to promote them to universe yet)

Reply:
> It would be awkward if our Makefile were in a ./ci sub-directory

Don't want to be nitpicky here -- just putting myself in the user's shoes. I chose `ci/` (actually @samalba did but I would have done the same) because contrary to a Makefile, the plan can use a bunch of different files and the filename is not standardized (not like Makefile, .gitignore, etc).

Same reason why the test infra scripts are in `infra/`, GitHub stuff is in `.github` and so on.

I agree with moving this to a top-level file because it makes the experience better, but as completely external user, it wouldn't have been my first choice (that's not how we manage other tools/scripts)

Reply:
I agree there is more DX subtlety to discuss here. On the one hand we will need a nice way for `dagger do` to work out of the box at the root of a project - I believe that would be a big downgrade from eg. make otherwise. But as you said, it's not clear how to do that without creating other awkwardness: no standard filename, could be a bunch of .cue files just lying around, etc.

Happy to discuss that topic further, anytime.